pandas/core/algorithms.py

```diff
@@ -642,6 +645,13 @@ def factorize(values, sort=False, order=None, na_sentinel=-1, size_hint=None):
                                         na_sentinel=na_sentinel,
                                         assume_unique=True)
 
+    if factor_nas and (labels == na_sentinel).any():
+        new_na_sentinel = len(uniques)
+        # Append the np.nan
+        uniques = np.resize(uniques, len(uniques) + 1)
+        uniques[-1] = np.nan
+        labels = np.where(labels == na_sentinel, new_na_sentinel, labels)
+
```

Comment:
Given that you're using broadcasting and `NumPy`, I don't think we should have any performance issues, but I wonder if we should still add a benchmark test anyway.